components/brave_wallet/browser/cardano/cardano_knapsack_solver.cc

```diff
@@ -0,0 +1,162 @@
+/* Copyright (c) 2025 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/cardano/cardano_knapsack_solver.h"
+
+#include <algorithm>
+#include <utility>
+
+#include "base/rand_util.h"
+#include "base/types/expected.h"
+#include "brave/components/brave_wallet/browser/cardano/cardano_serializer.h"
+#include "components/grit/brave_components_strings.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace brave_wallet {
+
+namespace {
+
+constexpr int kCardanoKnapsackSolverIterations = 1000;
+
+uint64_t ApplyFeeRate(uint64_t min_fee_coefficient,
+                      uint64_t min_fee_constant,
+                      uint32_t tx_size) {
+  return tx_size * min_fee_coefficient + min_fee_constant;
+}
+
+}  // namespace
+
+CardanoKnapsackSolver::CardanoKnapsackSolver(
+    CardanoTransaction base_transaction,
+    uint64_t min_fee_coefficient,
+    uint64_t min_fee_constant,
+    std::vector<CardanoTransaction::TxInput> inputs)
+    : base_transaction_(std::move(base_transaction)),
+      min_fee_coefficient_(min_fee_coefficient),
+      min_fee_constant_(min_fee_constant),
+      inputs_(std::move(inputs)) {}
+CardanoKnapsackSolver::~CardanoKnapsackSolver() = default;
+
+// static
+uint64_t CardanoKnapsackSolver::GetCostOfChangeOutput(
+    const CardanoTransaction& tx,
+    uint64_t min_fee_coefficient,
+    uint64_t min_fee_constant) {
+  CHECK(tx.ChangeOutput());
+
+  auto fee_with_change =
+      ApplyFeeRate(min_fee_coefficient, min_fee_constant,
+                   CardanoSerializer::CalcTransactionSize(tx));
+
+  auto tx_without_change = tx;
+  tx_without_change.ClearChangeOutput();
+  auto fee_without_change =
+      ApplyFeeRate(min_fee_coefficient, min_fee_constant,
+                   CardanoSerializer::CalcTransactionSize(tx_without_change));
+
+  if (fee_with_change > fee_without_change) {
+    return fee_with_change - fee_without_change;
+  }
+  return 0;
+}
+
+void CardanoKnapsackSolver::SolveForTransaction(
+    const CardanoTransaction& transaction,
+    std::optional<CardanoTransaction>& solution) {
+  // Don't create transaction if output's amount appears to be less than this
+  // threshold. Cost of spending such output should not be higher than its
+  // value.
+  uint64_t dust_output_threshold =
+      transaction.ChangeOutput()
+          ? GetCostOfChangeOutput(transaction, min_fee_coefficient_,
+                                  min_fee_constant_)
+          : 0;
+
+  for (int i = 0; i < kCardanoKnapsackSolverIterations; ++i) {
+    std::vector<bool> picked_groups(inputs_.size(), false);
```

Comment:
it looks like we should be hoisting this vector declaration outside this loop, rather than instantiating it 1k times (although there's always a very remote chance the compiler is already doing this)

Reply:
Moved allocation out of loop